[Frontend] Implement robust video frame recovery for corrupted videos

vSeamar opened the pull request:
<!-- markdownlint-disable -->

## Purpose

This PR implements **optional** video frame recovery functionality for vLLM's multimodal video processing using a dynamic window forward-scan approach.

### Key Changes:

- **Video Frame Recovery System (Optional)**: Added intelligent recovery logic that attempts to recover missing/broken video frames when sequential reading fails. **Disabled by default** - only activates when `frame_recovery=True`.

- **Dynamic Window Forward-Scan Algorithm**: When a target frame fails to load, the system continues scanning forward and uses the next successfully grabbed frame (before the next target frame) as a substitute.

- **Comprehensive Testing**: Added 3 new unit/regression tests covering recovery functionality with simulated failures, real corrupted files, and both video backends.

### Technical Details:

- **Single-Pass Algorithm**: The recovery system scans through frames linearly, tracking failed frames in a queue. When a valid frame is found within the "recovery window" (between a failed frame and the next target), it's used to recover the oldest failure.

- **Recovery Window Concept**: Each failed frame has a window extending to the next target frame. This ensures recovered frames maintain temporal proximity to their intended position.

- **Temporal Safety**: Recovered frames are always from *after* the failed frame (forward scan), ensuring the model never sees temporal inversion.

- **Backward Compatible**: Default behavior unchanged - `frame_recovery=False` (default) disables recovery entirely, maintaining existing performance and behavior.

- **Opt-in Enhancement**: Users can enable recovery via CLI:

  ```bash
  --media-io-kwargs '{"video": {"frame_recovery": true}}'
  ```

- **Backend Support**: Recovery works for both `opencv` and `opencv_dynamic` video backends when enabled.

- **Logging**: Recovery attempts and results are logged at INFO level for monitoring.

### Algorithm Overview:

```
Target frames:    [0]          [10]         [20]         [30]
                   ↓            ↓            ↓            ↓
Video timeline:  [0][1][2]...[10][11]...[19][20]...[30]...
                              ╳
                           (failed)
                              
Recovery window for frame 10: [10 → 20)
Frame 11 grabbed successfully → used to recover frame 10
```

## Test Plan

### Unit Tests Added:

- `test_video_recovery_simulated_failures`: Mocks `cv2.VideoCapture.grab()` to simulate failures on specific frames, verifies recovery successfully loads frames and preserves temporal order
- `test_video_recovery_with_corrupted_file`: Tests recovery using real corrupted video asset
- `test_video_recovery_dynamic_backend`: Tests recovery in dynamic frame sampling backend

### Test Coverage:

- Simulated frame grab failures
- Real corrupted video files
- Both video backend types (`opencv`, `opencv_dynamic`)
- Temporal ordering preservation
- Recovery metadata correctness

## Test Result

```
tests/multimodal/test_video.py - 14 passed
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

vSeamar committed:
fix mypy signature compatibility

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Fix NamedTemporaryFile parameter: delete_on_close -> delete

The parameter 'delete_on_close' doesn't exist in Python's NamedTemporaryFile.
Changed to 'delete=False' which is the correct parameter to prevent
automatic deletion when the file is closed.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Add comprehensive tests for video frame recovery functionality

- test_video_recovery_functionality: Tests successful recovery when sequential reading fails
- test_video_recovery_disabled: Tests that recovery is skipped when recovery_offset=0
- test_video_recovery_dynamic_backend: Tests recovery in dynamic frame sampling backend
- test_video_recovery_failure_logging: Tests proper logging when recovery fails

These tests ensure the recovery functionality works correctly and handles
edge cases like complete recovery failure and disabled recovery.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
test linting

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Add validation for negative recovery_offset parameter

- Add ValueError validation in both OpenCV backends for recovery_offset < 0
- Add test case to verify negative recovery_offset raises appropriate error
- Prevents unexpected behavior and provides clear error messages for invalid input

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Remove log assertions from recovery tests

Removed caplog dependencies and log message assertions from recovery tests:
- test_video_recovery_functionality
- test_video_recovery_disabled
- test_video_recovery_dynamic_backend
- test_video_recovery_failure_logging

Tests now focus on functional behavior verification without logging assertions,
avoiding issues with vLLM's custom logger configuration.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Add documentation for video frame recovery feature

Document the optional recovery_offset parameter in multimodal inputs guide,
including usage examples and parameter explanation.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Update video recovery documentation wording

Clarify that recovery works with video formats supported by OpenCV,
providing examples of common formats like MP4.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar commented on vllm/benchmarks/datasets.py:
```diff
@@ -862,7 +862,7 @@ def generate_synthetic_video(
         fps = 30  # frames per second
 
-        with NamedTemporaryFile(suffix=".mp4", delete_on_close=False) as temp_file:
+        with NamedTemporaryFile(suffix=".mp4", delete=False) as temp_file:
```
benchmark tests would fail on python 3.10.16 (which should be supported):
`pytest tests/benchmarks/test_random_multimodal_dataset_video.py -v`

This change now passes all the tests on 3.10.16 and 3.13.6.

vSeamar committed:
Add comments

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Fix markdown formatting in multimodal inputs documentation

Remove trailing spaces and add blank line before list items
to comply with markdownlint rules.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Fix test mock function to match updated _load_frames_with_seeking signature

Add missing existing_frames parameter to mock_seek_failure function
to match the optimized recovery method signature.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Optimize video frame recovery with multi-strategy approach

Implement high-performance recovery logic with three strategies:
- Cache Check: Zero I/O reuse of sequential frames
- Forward Scan: Single seek + grab for corruption gaps
- Backward Scan: Single seek + grab for truncation

Add existing_frames parameter to enable cache-based recovery.
Significantly reduces I/O operations and improves performance.

Signed-off-by: cmartinez <cmartinez@roblox.com>

Isotr0py commented on vllm/multimodal/video.py:
```diff
@@ -63,12 +63,160 @@ def load_bytes(
+        1. Cache Check (Zero I/O): Reuse frames from sequential read if close enough.
+        2. Forward Scan (Low I/O): Seek once to target+1, grab forward (handles gaps).
+        3. Backward Scan (Med I/O): Seek to target-offset, grab forward (truncation).
+        """
```
Hmm, I feel current recovery implementation is a bit complicated. I wonder if we can recover frames from loaded numpy ndarray directly?

For example, given that `num_expected_framess=6` and `valid_frames=[0, 1, 3, 5]` with `broken_frames = [2, 4]`, we can know that frames used to fix are `[1, 3]` or `[3, 5]`. Then we recover frames with numpy like: 
```python3
num_expected_frames = 6
valid_frames = [0, 1, 3, 5]
invalid_frames = [2, 4]
...  # calculate frames_idx used to fix here
frames_used_to_recover = [1, 3]   # backward
frames_used_to_recover = [3, 5]   # forward
frames[invalid_frames] = frames[frames_used_to_recover]
```
I think this can avoid expensive OpenCV I/O operation, WDYT?

Isotr0py replied:
I think we can try to add forward scan with a simplified implementation in this PR first, then consider remained strategy. Because we will need to refactor the loading logic before supporting all three or more recovery strategies, otherwise the video loading logic will become too complicated. 🤔 

We can simply implement forward scan like this, if we found broken frame during loading, we can keep trying loading next possible frame used to recovery until finding a valid ones:
```python3
        i = 0
        valid_frame_indices = []
        failed_frames_idx = []
        for idx in range(max_frame_idx + 1):
            is_target_frame = idx in frame_indices
            ok = cap.grab()
            if not ok:
                # Frame is broken/unreadable, log warning
                if is_target_frame:
                    failed_frames_idx.append(idx)
                continue
            if is_target_frame or (can_recover := can_use_to_recover_frame(idx, failed_frames_idx, ...)):
                ret, frame = cap.retrieve()
                if ret:
                    frames[i] = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
                    valid_frame_indices.append(idx)
                    i += 1
                    if can_recover:
                        failed_frames_idx.pop(0)
                elif is_target_frame:
                    failed_frames_idx.append(idx)
```
Here, `can_use_to_recover_frame` should be a util function to check if current frame available for recovery. WDYT?

vSeamar committed:
Implement dynamic window forward-scan frame recovery

Replace complex multi-strategy recovery with simpler forward-scan approach:
- When a target frame fails, use the next valid frame within the window
- Dynamic window extends until the next target frame
- Simple boolean frame_recovery parameter instead of recovery_offset

Changes:
- Add _read_frames_with_recovery method with dynamic window logic
- Update load_bytes to accept frame_recovery boolean parameter
- Add tests for recovery with corrupted files
- Update documentation with frame_recovery usage

The forward-scan approach is simpler and more predictable than seeking,
while still providing recovery for mid-video corruption and truncation.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Remove redundant frame_recovery parameter from _read_frames_with_recovery

The method is only called when frame_recovery=True, so the parameter
check inside was unnecessary. Simplified the code by removing it.

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Add integration test for frame recovery with simulated failures

Uses simulate_corruption.mp4 asset and mocks VideoCapture.grab() to
fail on specific frames, then verifies:
- Without recovery: fewer frames due to simulated failures
- With recovery: more frames recovered using next valid frame
- Temporal order is preserved in the output

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Clean up comments and change recovery log level to info

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

Signed-off-by: vSeamar <cmartinez@roblox.com>

vSeamar committed:
Refactor recovery logic to cleaner single-pass approach

- Add _can_use_for_recovery() helper function
- Simplify main loop: single append per frame instead of double
- Remove unnecessary sorting step (frames now read in order)
- Fix issue where same frame could be appended twice

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Add type annotation for cv2.VideoCapture parameter to fix docs build

Signed-off-by: cmartinez <cmartinez@roblox.com>

Isotr0py reviewed (approved):
LGTM now, thanks for your patience! Just leave some nits.

vSeamar committed:
Improve video recovery tests with corrupted.mp4

- Enhanced test_video_recovery_with_corrupted_file to properly validate
  real H.264 reference frame chain corruption behavior
- Updated test_video_recovery_dynamic_backend to compare with/without
  recovery instead of only testing one mode
- Added detailed docstrings explaining the difference between real codec
  corruption (unrecoverable) vs simulated transient failures (recoverable)
- Both tests now provide complementary coverage: unit test for algorithm
  correctness and integration test for real-world robustness
- Fixed line length issue in video.py assert statement

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Consolidate recovery tests to use corrupted.mp4 only

- Remove simulate_corruption.mp4 (2.1 MB) as it's no longer needed
- Update test_video_recovery_simulated_failures to use corrupted.mp4
  with mocked failures on frames 3 and 10 (plus real corruption at 17)
- Update test_video_recovery_with_corrupted_file to use sparse sampling
  (num_frames=8) which makes frame 17 a target with recovery window [17, 21)
- Both tests now verify recovery actually WORKS with real corruption
- Reduces repo size and addresses code review feedback about duplicate videos

Signed-off-by: cmartinez <cmartinez@roblox.com>

vSeamar commented on tests/multimodal/assets/simulate_corruption.mp4:
Done! We now use only corrupted.mp4 for all recovery tests (refactored to keep same testing functionality bu with only corrupted.mp4). The duplicate video has been removed.

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar commented:
@Isotr0py 
buildkite/ci/pr seems to not get picked up even after a few hours. Is this normal?

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar committed:
Merge branch 'main' into cm/frameRecovery

vSeamar commented:
@Isotr0py Could I get some guidance on how the merging process is? I keep updating branch with main but the checks take very long and I usually have to update it again before they finish. 

Or since you enabled auto-merge do I not have to do anything anymore?

Isotr0py commented:
> Or since you enabled auto-merge do I not have to do anything anymore?

Usually it can automatically merge as long as there is no conflict with main, even if it's out of date. (Our CI takes a quite long time to run 😅 )
